File: relay/provisioning_cipher.js
// vim: ts=4:sw=4:expandtab

'use strict';

const assert = require('assert');
const crypto = require('crypto');
const libsignal = require('libsignal');
const protobufs = require('./protobufs');


class ProvisioningCipher {

    async decrypt(provisionEnvelope) {
        const masterEphemeral = provisionEnvelope.publicKey;
        const message = provisionEnvelope.body;
        if (message[0] !== 1) {
            throw new Error("Bad version number on ProvisioningMessage");
        }
        const iv = message.slice(1, 16 + 1);
        const mac = message.slice(message.byteLength - 32, message.byteLength);
        const ivAndCiphertext = message.slice(0, message.byteLength - 32);
        const ciphertext = message.slice(16 + 1, message.byteLength - 32);
        const ecRes = libsignal.crypto.calculateAgreement(masterEphemeral, this.keyPair.privKey);
        const keys = await libsignal.crypto.HKDF(ecRes, Buffer.alloc(32),
            Buffer.from("TextSecure Provisioning Message"));
        await libsignal.crypto.verifyMAC(ivAndCiphertext, keys[1], mac, 32);
        const plaintext = await libsignal.crypto.decrypt(keys[0], ciphertext, iv);
        const provisionMessage = protobufs.ProvisionMessage.decode(plaintext);
        const privKey = provisionMessage.identityKeyPrivate;
        return {
            identityKeyPair: libsignal.crypto.createKeyPair(privKey),
            addr: provisionMessage.addr,
            provisioningCode: provisionMessage.provisioningCode,
            userAgent: provisionMessage.userAgent
        };
    }

    async encrypt(theirPublicKey, message) {
        assert(theirPublicKey instanceof Buffer);
        const ourKeyPair = libsignal.crypto.generateKeyPair();
        const sharedSecret = libsignal.crypto.calculateAgreement(theirPublicKey,
                                                                 ourKeyPair.privKey);
        const derivedSecret = await libsignal.crypto.HKDF(sharedSecret, Buffer.alloc(32),
            Buffer.from("TextSecure Provisioning Message"));
        const ivLen = 16;
        const macLen = 32;
        const iv = crypto.randomBytes(ivLen);
        const encryptedMsg = await libsignal.crypto.encrypt(derivedSecret[0], message /* XXX validate is Buffer / right */, iv);
        const msgLen = encryptedMsg.byteLength;

        const data = new Uint8Array(1 + ivLen + msgLen);
        data[0] = 1;  // Version
        data.set(iv, 1);
        data.set(new Uint8Array(encryptedMsg), 1 + ivLen);
        const mac = await libsignal.crypto.calculateMAC(derivedSecret[1], data.buffer);
        const pEnvelope = new protobufs.ProvisionEnvelope();
        pEnvelope.body = new Uint8Array(data.byteLength + macLen);
        pEnvelope.body.set(data, 0);
        pEnvelope.body.set(new Uint8Array(mac), data.byteLength);
        pEnvelope.publicKey = ourKeyPair.pubKey;
        return pEnvelope;
    }

    getPublicKey() {
        if (!this.keyPair) {
            this.keyPair = libsignal.crypto.generateKeyPair();
        }
        return this.keyPair.pubKey;
    }
}

module.exports = ProvisioningCipher;

File: relay/provisioning_cipher.py
"""
Device onboarding protocol
"""

from . import protobufs
from axolotl.ecc.curve import Curve
from axolotl.ecc.eckeypair import ECKeyPair
from axolotl.ecc.djbec import DjbECPublicKey, DjbECPrivateKey
from axolotl.kdf.hkdf import HKDF
from axolotl.protocol.whispermessage import WhisperMessage
from axolotl.sessioncipher import AESCipher


class ProvisioningCipher(object):

    async def decrypt(self, provisionEnvelope):
        masterEphemeral = provisionEnvelope.publicKey
        message = provisionEnvelope.body
        if message[0] != 1:
            raise ValueError("Bad version number on ProvisioningMessage")
        iv = message[1:16 + 1]
        mac = message[-32:]
        ivAndCiphertext = message[:-32]
        ciphertext = message[16 + 1:-32]
        ecRes = Curve.calculateAgreement(masterEphemeral, self.keypair.privateKey)
        keys = HKDF.createFor(3).deriveSecrets(ecRes, "TextSecure Provisioning Message", 64)
        wm = WhisperMessage()
        wm.verifyMAC(1, ivAndCiphertext, keys[1], mac)
        plaintext = AESCipher(keys[0], iv).decrypt(ciphertext)
        provisionMessage = protobufs.ProvisionMessage.decode(plaintext)
        privateKey = provisionMessage.identityKeyPrivate
        publicKey = Curve.generatePublicKey(privateKey)
        return {
            "identityKeyPair": ECKeyPair(DjbECPublicKey(publicKey),
                                         DjbECPrivateKey(privateKey)),
            "addr": provisionMessage.addr,
            "provisioningCode": provisionMessage.provisioningCode,
            "userAgent": provisionMessage.userAgent
        }

    async def encrypt(self, theirPublicKey, message):
        ourKeyPair = Curve.generateKeyPair()
        sharedSecret = Curve.calculateAgreement(theirPublicKey,
                                                ourKeyPair.privateKey)
        '''
        derivedSecret = await libsignal.crypto.HKDF(sharedSecret, Buffer.alloc(32),
            Buffer.from("TextSecure Provisioning Message"))
        ivLen = 16
        macLen = 32
        iv = crypto.randomBytes(ivLen)
        encryptedMsg = await libsignal.crypto.encrypt(derivedSecret[0], message /* XXX validate is Buffer / right */, iv)
        msgLen = encryptedMsg.byteLength

        data = Uint8Array(1 + ivLen + msgLen)
        data[0] = 1;  // Version
        data.set(iv, 1)
        data.set(Uint8Array(encryptedMsg), 1 + ivLen)
        mac = await libsignal.crypto.calculateMAC(derivedSecret[1], data.buffer)
        pEnvelope = protobufs.ProvisionEnvelope()
        pEnvelope.body = Uint8Array(data.byteLength + macLen)
        pEnvelope.body.set(data, 0)
        pEnvelope.body.set(Uint8Array(mac), data.byteLength)
        pEnvelope.publicKey = ourKeyPair.publicKey
        return pEnvelope
        '''

    def get_publickey(self):
        if not self.keypair:
            self.keypair = Curve.generateKeyPair()
        return self.keypair.publicKey
